src/context/blocked.rs

```diff
@@ -1,18 +1,177 @@
-use std::ptr;
+use crate::redismodule::{AUTH_HANDLED, AUTH_NOT_HANDLED};
+use crate::{raw, Context, ValkeyError, ValkeyString};
+use std::os::raw::{c_int, c_void};
 
-use crate::raw;
-use crate::Context;
+// Callback types for handling blocked client operations
+#[derive(Debug)]
+pub enum ReplyCallback<T> {
+    Auth(fn(&Context, ValkeyString, ValkeyString, T) -> Result<c_int, ValkeyError>),
+}
+
+// Callback type for freeing private data associated with a blocked client
+type FreePrivDataCallback<T> = fn(&Context, T);
 
-pub struct BlockedClient {
+pub struct BlockedClient<T = ()> {
     pub(crate) inner: *mut raw::RedisModuleBlockedClient,
+    reply_callback: Option<ReplyCallback<T>>,
+    free_priv_callback: Option<FreePrivDataCallback<T>>,
+    private_data: Option<Box<T>>,
+}
+
+#[allow(dead_code)]
+unsafe extern "C" fn auth_reply_wrapper<T: 'static + Clone>(
+    ctx: *mut raw::RedisModuleCtx,
+    username: *mut raw::RedisModuleString,
+    password: *mut raw::RedisModuleString,
+    err: *mut *mut raw::RedisModuleString,
+) -> c_int {
+    let context = Context::new(ctx);
+    let ctx_ptr = std::ptr::NonNull::new_unchecked(ctx);
+    let username = ValkeyString::new(Some(ctx_ptr), username);
+    let password = ValkeyString::new(Some(ctx_ptr), password);
```

Comment:
There should be an API of the `ValkeyString` struct that accepts a `*mut raw::RedisModuleCtx` directly so that you don't have to box this structure

The other potential issue of using `new` is that it calls string_retain_string which increments a ref counter and this causes a leak of ValkeyString in the core:
```
    pub fn new(
        ctx: Option<NonNull<raw::RedisModuleCtx>>,
        inner: *mut raw::RedisModuleString,
    ) -> Self {
        let ctx = ctx.map_or(std::ptr::null_mut(), |v| v.as_ptr());
        raw::string_retain_string(ctx, inner);
        Self { ctx, inner }
    }
```